### PR TITLE
K8SPXC-749 add possibility of changing all probes options via CR 

### PR DESCRIPTION
**Pull request by hors**
[K8SPXC-749](https://jira.percona.com/browse/K8SPXC-749) [Powered by Pull Request Badge](https://pullrequestbadge.com/?utm_medium=github&utm_source=percona&utm_campaign=badge_info)

### CODE REVIEW

File: pkg/apis/pxc/v1/pxc_types.go [hunk 1]
```diff
@@ -718,6 +721,74 @@ const (
 	minSafeProxySize = 2
 )
 
+func (cr *PerconaXtraDBCluster) setProbesDefaults() {
+	if cr.Spec.PXC != nil {
+		if cr.Spec.PXC.LivenessInitialDelaySeconds != nil {
+			cr.Spec.PXC.LivenessProbes.InitialDelaySeconds = *cr.Spec.PXC.LivenessInitialDelaySeconds
+		} else if cr.Spec.PXC.LivenessProbes.InitialDelaySeconds == 0 {
+			cr.Spec.PXC.LivenessProbes.InitialDelaySeconds = 300
+		}
```
Comment (delgod) [inline pkg/apis/pxc/v1/pxc_types.go:730]: just Q: why it is needed to convert 0 to 300?

File: pkg/pxc/app/statefulset/haproxy.go [hunk 2]
```diff
@@ -166,6 +166,25 @@ func (c *HAProxy) AppContainer(spec *api.PodSpec, secrets string, cr *api.Percon
...
+			{
+				Name:  "LIVENESS_CHECK_TIMEOUT",
+				Value: fmt.Sprint(cr.Spec.HAProxy.LivenessProbes.TimeoutSeconds),
+			},
+			{
+				Name:  "READINESS_CHECK_TIMEOUT",
+				Value: fmt.Sprint(cr.Spec.HAProxy.ReadinessProbes.TimeoutSeconds),
+			},
+		}
+		appc.Env = append(appc.Env, probsEnvs...)
```
Comment (delgod) [inline pkg/pxc/app/statefulset/haproxy.go:187]: we should add them only if version >= 1.10

**Review by delgod — CHANGES_REQUESTED** [review]

File: pkg/pxc/app/statefulset/haproxy.go (hunk 2 above)
Comment (hors) [reply pkg/pxc/app/statefulset/haproxy.go:187]: I want to jump to the outgoing train but 1.10 is ok for me too.

File: pkg/apis/pxc/v1/pxc_types.go (hunk 1 above)
Comment (hors) [reply pkg/apis/pxc/v1/pxc_types.go:730]: E.g. we do not have `spec.pxc.livenessProbe.initialDelaySeconds` option at all in CR we have '0' here but need to change it to our custom default value (300). That is why we have such checks.

File: deploy/cr.yaml [hunk 3]
```diff
@@ -226,8 +250,8 @@ spec:
     gracePeriod: 30
 #   loadBalancerSourceRanges:
 #     - 10.0.0.0/8
-#   serviceAnnotations:
-#     service.beta.kubernetes.io/aws-load-balancer-backend-protocol: http
+#    serviceAnnotations:
+#      service.beta.kubernetes.io/aws-load-balancer-backend-protocol: http
```
[the remark this answers is not in the document]
Comment (hors) [reply deploy/cr.yaml:254]: It is fix for our CR, but thanks I see additional one.

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| big-data | passed |
| storage | failed |
| upgrade-haproxy | failed |
| scheduled-backup | failed |
| scaling | passed |
| recreate | passed |
| haproxy | passed |
| scaling-proxysql | passed |
| init-deploy | failed |
| self-healing | passed |
| restore-to-encrypted-cluster | passed |
| security-context | failed |
| limits | failed |
| demand-backup | passed |
| monitoring-2-0 | failed |
| upgrade-proxysql | failed |
| self-healing-chaos | passed |
| affinity | failed |
| self-healing-advanced | passed |
| one-pod | failed |
| pitr | passed |
| auto-tuning | passed |
| self-healing-advanced-chaos | failed |
| smart-update | passed |
| proxysql-sidecar-res-limits | passed |
| operator-self-healing | passed |
| users | passed |
| operator-self-healing-chaos | failed |
| tls-issue-self | passed |
| tls-issue-cert-manager | passed |
| demand-backup-encrypted-with-tls | passed |
| tls-issue-cert-manager-ref | passed |
| validation-hook | passed |
| upgrade-consistency | failed |

commit: https://github.com/percona/percona-xtradb-cluster-operator/pull/946/commits/26d6471b76ef9e29336906280f5e61cd609a39c4
image: `perconalab/percona-xtradb-cluster-operator:PR-946-26d6471b`

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| scheduled-backup | passed |
| big-data | passed |
| haproxy | failed |
| storage | passed |
| scaling | passed |
| recreate | passed |
| upgrade-haproxy | passed |
| self-healing | passed |
| scaling-proxysql | passed |
| security-context | passed |
| restore-to-encrypted-cluster | passed |
| upgrade-proxysql | passed |
| init-deploy | passed |
| limits | passed |
| smart-update | passed |
| demand-backup | passed |
| self-healing-chaos | passed |
| monitoring-2-0 | passed |
| self-healing-advanced | passed |
| affinity | passed |
| one-pod | passed |
| auto-tuning | passed |
| self-healing-advanced-chaos | passed |
| pitr | passed |
| proxysql-sidecar-res-limits | passed |
| upgrade-consistency | passed |
| users | passed |
| operator-self-healing | passed |
| operator-self-healing-chaos | passed |
| tls-issue-self | passed |
| demand-backup-encrypted-with-tls | passed |
| tls-issue-cert-manager | passed |
| tls-issue-cert-manager-ref | passed |
| validation-hook | passed |

commit: https://github.com/percona/percona-xtradb-cluster-operator/pull/946/commits/b73ed94cea9d03de14f0a8d13a2af2f30fcba10c
image: `perconalab/percona-xtradb-cluster-operator:PR-946-b73ed94c`

File: pkg/apis/pxc/v1/pxc_types.go [hunk 4]
```diff
@@ -718,6 +721,74 @@ const (
...
+		if cr.Spec.HAProxy.LivenessProbes.TimeoutSeconds == 0 {
+			cr.Spec.HAProxy.LivenessProbes.TimeoutSeconds = 5
+		}
+		if cr.Spec.HAProxy.LivenessProbes.FailureThreshold == 0 {
+			cr.Spec.HAProxy.LivenessProbes.FailureThreshold = 4
+		}
+		if cr.Spec.HAProxy.LivenessProbes.PeriodSeconds == 0 {
+			cr.Spec.HAProxy.LivenessProbes.PeriodSeconds = 30
+		}
+		if cr.Spec.HAProxy.LivenessProbes.SuccessThreshold > 1 {
```
Comment (delgod) [inline pkg/apis/pxc/v1/pxc_types.go:785]: literally, allowed values 0 or 1. why?

File: pkg/pxc/app/statefulset/haproxy.go (hunk 2 above)
Comment (delgod) [reply pkg/pxc/app/statefulset/haproxy.go:189]: ok, lets do for 1.9

**Review by delgod — CHANGES_REQUESTED** [review]

File: pkg/apis/pxc/v1/pxc_types.go (hunk 4 above)
Comment (hors) [reply pkg/apis/pxc/v1/pxc_types.go:785]: please see description:
```
	// Minimum consecutive successes for the probe to be considered successful after having failed.
	// Defaults to 1. Must be 1 for liveness and startup. Minimum value is 1.
```

File: pkg/pxc/app/statefulset/haproxy.go (hunk 2 above)
Comment (hors) [reply pkg/pxc/app/statefulset/haproxy.go:188]: ok

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| recreate | passed |
| big-data | passed |
| haproxy | passed |
| upgrade-haproxy | failed |
| scheduled-backup | passed |
| scaling | passed |
| storage | failed |
| init-deploy | failed |
| scaling-proxysql | passed |
| self-healing | passed |
| restore-to-encrypted-cluster | passed |
| security-context | failed |
| limits | failed |
| demand-backup | passed |
| monitoring-2-0 | failed |
| self-healing-chaos | passed |
| upgrade-proxysql | failed |
| affinity | failed |
| self-healing-advanced | passed |
| one-pod | failed |
| pitr | failed |
| auto-tuning | passed |
| self-healing-advanced-chaos | passed |
| proxysql-sidecar-res-limits | passed |
| users | failed |
| smart-update | passed |
| operator-self-healing | passed |
| operator-self-healing-chaos | passed |
| upgrade-consistency | failed |
| tls-issue-self | passed |
| demand-backup-encrypted-with-tls | passed |
| tls-issue-cert-manager | passed |
| tls-issue-cert-manager-ref | passed |
| validation-hook | passed |

commit: https://github.com/percona/percona-xtradb-cluster-operator/pull/946/commits/761931c0a9d7b761d305cffad39b052b9914cc89
image: `perconalab/percona-xtradb-cluster-operator:PR-946-761931c0`

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| storage | failed |
| haproxy | passed |
| recreate | passed |
| big-data | passed |
| scaling | passed |
| scheduled-backup | passed |
| upgrade-haproxy | failed |
| init-deploy | failed |
| self-healing | passed |
| scaling-proxysql | passed |
| security-context | failed |
| restore-to-encrypted-cluster | passed |
| limits | failed |
| self-healing-chaos | passed |
| monitoring-2-0 | failed |
| demand-backup | passed |
| upgrade-proxysql | failed |
| self-healing-advanced | passed |
| affinity | failed |
| self-healing-advanced-chaos | passed |
| one-pod | failed |
| auto-tuning | passed |
| pitr | passed |
| operator-self-healing | passed |
| smart-update | passed |
| proxysql-sidecar-res-limits | passed |
| users | failed |
| operator-self-healing-chaos | passed |
| demand-backup-encrypted-with-tls | passed |
| upgrade-consistency | failed |
| tls-issue-self | passed |
| tls-issue-cert-manager | passed |
| tls-issue-cert-manager-ref | passed |
| validation-hook | passed |

commit: https://github.com/percona/percona-xtradb-cluster-operator/pull/946/commits/cf78cf1b3fd09cf703a8002aecdd11c2280888db
image: `perconalab/percona-xtradb-cluster-operator:PR-946-cf78cf1b`

File: pkg/apis/pxc/v1/pxc_types.go [hunk 5]
```diff
@@ -718,6 +721,74 @@ const (
...
+		} else if cr.Spec.PXC.LivenessProbes.InitialDelaySeconds == 0 {
+			cr.Spec.PXC.LivenessProbes.InitialDelaySeconds = 300
+		}
+
+		if cr.Spec.PXC.LivenessProbes.TimeoutSeconds == 0 {
+			cr.Spec.PXC.LivenessProbes.TimeoutSeconds = 5
+		}
+		if cr.Spec.PXC.LivenessProbes.SuccessThreshold > 1 || cr.Spec.PXC.LivenessProbes.SuccessThreshold == 0 {
+			cr.Spec.PXC.LivenessProbes.SuccessThreshold = 1
+		}
```
Comment (delgod) [inline pkg/apis/pxc/v1/pxc_types.go:737]: logic have no sense, just put `1` always
```suggestion
		cr.Spec.PXC.LivenessProbes.SuccessThreshold = 1
```
the same for HAProxy

File: pkg/apis/pxc/v1/pxc_types.go [hunk 6]
```diff
@@ -718,6 +721,74 @@ const (
...
+
+		if cr.Spec.PXC.ReadinessInitialDelaySeconds != nil {
+			cr.Spec.PXC.ReadinessProbes.InitialDelaySeconds = *cr.Spec.PXC.ReadinessInitialDelaySeconds
+		} else if cr.Spec.PXC.ReadinessProbes.InitialDelaySeconds == 0 {
+			cr.Spec.PXC.ReadinessProbes.InitialDelaySeconds = 15
```
Comment (delgod) [inline pkg/apis/pxc/v1/pxc_types.go:742]: I don't understand how code can reach this line
line 740 (not 742) will be executed `if cr.Spec.PXC.ReadinessProbes.InitialDelaySeconds == 0`

the same in HAProxy part

**Review by delgod — CHANGES_REQUESTED** [review]

File: pkg/apis/pxc/v1/pxc_types.go (hunk 6 above)
Comment (hors) [reply pkg/apis/pxc/v1/pxc_types.go:742]: We already have `readinessInitialDelaySeconds`  option in CR, and if this option is used, we set the old value  from it to new `InitialDelaySeconds` option.

File: pkg/apis/pxc/v1/pxc_types.go (hunk 5 above)
Comment (hors) [reply pkg/apis/pxc/v1/pxc_types.go:737]: oh, yes

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| haproxy | passed |
| recreate | passed |
| storage | passed |
| scaling | passed |
| upgrade-haproxy | failed |
| scheduled-backup | passed |
| big-data | passed |
| scaling-proxysql | passed |
| init-deploy | passed |
| self-healing | passed |
| restore-to-encrypted-cluster | passed |
| security-context | passed |
| limits | passed |
| self-healing-chaos | passed |
| monitoring-2-0 | failed |
| demand-backup | passed |
| self-healing-advanced | passed |
| upgrade-proxysql | passed |
| affinity | failed |
| self-healing-advanced-chaos | passed |
| pitr | passed |
| one-pod | passed |
| smart-update | passed |
| operator-self-healing | passed |
| auto-tuning | passed |
| operator-self-healing-chaos | passed |
| proxysql-sidecar-res-limits | passed |
| users | passed |
| demand-backup-encrypted-with-tls | passed |
| tls-issue-self | passed |
| upgrade-consistency | passed |
| tls-issue-cert-manager | passed |
| tls-issue-cert-manager-ref | passed |
| validation-hook | passed |

commit: https://github.com/percona/percona-xtradb-cluster-operator/pull/946/commits/39343d9765f86af242555acea1a23aa1df435da0
image: `perconalab/percona-xtradb-cluster-operator:PR-946-39343d97`

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |

commit: https://github.com/percona/percona-xtradb-cluster-operator/pull/946/commits/81c8b8ed15a41d459086507a51da9361925cb53d
image: `perconalab/percona-xtradb-cluster-operator:PR-946-81c8b8ed`

**Review by delgod — APPROVED** [review]

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| scaling | passed |
| haproxy | passed |
| recreate | passed |
| scheduled-backup | passed |
| storage | passed |
| upgrade-haproxy | failed |
| big-data | passed |
| scaling-proxysql | passed |
| self-healing | passed |
| init-deploy | passed |
| restore-to-encrypted-cluster | passed |
| security-context | passed |
| limits | passed |
| self-healing-chaos | passed |
| demand-backup | passed |
| monitoring-2-0 | passed |
| self-healing-advanced | passed |
| upgrade-proxysql | passed |
| affinity | passed |
| one-pod | passed |
| self-healing-advanced-chaos | passed |
| auto-tuning | passed |
| smart-update | passed |
| pitr | passed |
| proxysql-sidecar-res-limits | passed |
| operator-self-healing | passed |
| users | passed |
| operator-self-healing-chaos | passed |
| tls-issue-self | passed |
| demand-backup-encrypted-with-tls | passed |
| tls-issue-cert-manager | passed |
| tls-issue-cert-manager-ref | passed |
| upgrade-consistency | passed |
| validation-hook | passed |

commit: https://github.com/percona/percona-xtradb-cluster-operator/pull/946/commits/81c8b8ed15a41d459086507a51da9361925cb53d
image: `perconalab/percona-xtradb-cluster-operator:PR-946-81c8b8ed`

File: pkg/apis/pxc/v1/pxc_types.go [hunk 7]
```diff
@@ -718,6 +721,72 @@ const (
 	minSafeProxySize = 2
 )
 
+func (cr *PerconaXtraDBCluster) setProbesDefaults() {
+	if cr.Spec.PXC != nil {
```
Comment (qJkee) [inline pkg/apis/pxc/v1/pxc_types.go:725]: if we reach there, pxc is always not nil

File: pkg/apis/pxc/v1/pxc_types.go [hunk 8]
```diff
@@ -718,6 +721,72 @@ const (
...
+			cr.Spec.PXC.LivenessProbes.InitialDelaySeconds = *cr.Spec.PXC.LivenessInitialDelaySeconds
+		} else if cr.Spec.PXC.LivenessProbes.InitialDelaySeconds == 0 {
+			cr.Spec.PXC.LivenessProbes.InitialDelaySeconds = 300
+		}
+
+		if cr.Spec.PXC.LivenessProbes.TimeoutSeconds == 0 {
+			cr.Spec.PXC.LivenessProbes.TimeoutSeconds = 5
+		}
+
+		cr.Spec.PXC.LivenessProbes.SuccessThreshold = 1
```
Comment (qJkee) [inline pkg/apis/pxc/v1/pxc_types.go:736]: should we overwrite it? As i understand user can specify his own value

File: pkg/apis/pxc/v1/pxc_types.go [hunk 9]
```diff
@@ -718,6 +721,72 @@ const (
...
+			cr.Spec.HAProxy.LivenessProbes.TimeoutSeconds = 5
+		}
+		if cr.Spec.HAProxy.LivenessProbes.FailureThreshold == 0 {
+			cr.Spec.HAProxy.LivenessProbes.FailureThreshold = 4
+		}
+		if cr.Spec.HAProxy.LivenessProbes.PeriodSeconds == 0 {
+			cr.Spec.HAProxy.LivenessProbes.PeriodSeconds = 30
+		}
+
+		cr.Spec.HAProxy.LivenessProbes.SuccessThreshold = 1
```
Comment (qJkee) [inline pkg/apis/pxc/v1/pxc_types.go:785]: same, user can specify his own

File: pkg/apis/pxc/v1/pxc_types.go (hunk 8 above)
Comment (hors) [reply pkg/apis/pxc/v1/pxc_types.go:736]: yes, please see comment https://github.com/percona/percona-xtradb-cluster-operator/pull/946#discussion_r676372373

File: pkg/apis/pxc/v1/pxc_types.go (hunk 9 above)
Comment (hors) [reply pkg/apis/pxc/v1/pxc_types.go:785]: no, please see comment https://github.com/percona/percona-xtradb-cluster-operator/pull/946#discussion_r676372373

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| scaling | passed |
| haproxy | passed |
| storage | passed |
| recreate | passed |
| big-data | passed |
| scaling-proxysql | passed |
| self-healing | passed |
| init-deploy | passed |
| security-context | passed |
| restore-to-encrypted-cluster | passed |
| limits | passed |
| self-healing-chaos | passed |
| monitoring-2-0 | passed |
| demand-backup | passed |
| self-healing-advanced | passed |
| affinity | passed |
| one-pod | passed |
| auto-tuning | passed |
| self-healing-advanced-chaos | passed |
| proxysql-sidecar-res-limits | passed |
| pitr | failed |
| operator-self-healing | passed |
| users | passed |
| operator-self-healing-chaos | passed |
| tls-issue-self | passed |
| tls-issue-cert-manager | passed |
| tls-issue-cert-manager-ref | passed |
| validation-hook | passed |

commit: https://github.com/percona/percona-xtradb-cluster-operator/pull/946/commits/27832395488835bb668b850ac3bcb8fa369a0b89
image: `perconalab/percona-xtradb-cluster-operator:PR-946-27832395`

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| big-data | passed |
| upgrade-haproxy | passed |
| scheduled-backup | passed |
| storage | passed |
| scaling | passed |
| haproxy | passed |
| recreate | passed |
| self-healing | passed |
| scaling-proxysql | passed |
| init-deploy | passed |
| security-context | passed |
| restore-to-encrypted-cluster | passed |
| upgrade-proxysql | passed |
| limits | passed |
| monitoring-2-0 | passed |
| demand-backup | passed |
| smart-update | passed |
| self-healing-chaos | passed |
| affinity | passed |
| one-pod | passed |
| self-healing-advanced | passed |
| auto-tuning | passed |
| proxysql-sidecar-res-limits | passed |
| pitr | passed |
| users | failed |
| self-healing-advanced-chaos | passed |
| operator-self-healing | passed |
| operator-self-healing-chaos | passed |
| demand-backup-encrypted-with-tls | failed |
| upgrade-consistency | passed |
| tls-issue-self | passed |
| tls-issue-cert-manager | passed |
| tls-issue-cert-manager-ref | passed |
| validation-hook | passed |

commit: https://github.com/percona/percona-xtradb-cluster-operator/pull/946/commits/4a7db60402529e966db78a96ac64be5b61f48e82
image: `perconalab/percona-xtradb-cluster-operator:PR-946-4a7db604`

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| upgrade-haproxy | passed |
| recreate | passed |
| big-data | passed |
| storage | passed |
| scheduled-backup | passed |
| self-healing | passed |
| restore-to-encrypted-cluster | passed |
| self-healing-chaos | passed |
| demand-backup | passed |
| haproxy | passed |
| scaling | passed |
| self-healing-advanced | passed |
| pitr | passed |
| init-deploy | passed |
| scaling-proxysql | passed |
| self-healing-advanced-chaos | passed |
| demand-backup-encrypted-with-tls | passed |
| limits | passed |
| security-context | passed |
| operator-self-healing | passed |
| monitoring-2-0 | passed |
| operator-self-healing-chaos | passed |
| affinity | passed |
| one-pod | passed |
| auto-tuning | passed |
| proxysql-sidecar-res-limits | passed |
| users | passed |
| upgrade-proxysql | passed |
| smart-update | passed |
| upgrade-consistency | passed |
| tls-issue-self | passed |
| tls-issue-cert-manager | passed |
| tls-issue-cert-manager-ref | passed |
| validation-hook | passed |

commit: https://github.com/percona/percona-xtradb-cluster-operator/pull/946/commits/4a7db60402529e966db78a96ac64be5b61f48e82
image: `perconalab/percona-xtradb-cluster-operator:PR-946-4a7db604`

**Review by qJkee — APPROVED** [review]

**Review by delgod — APPROVED** [review]

**Pull request merged**